src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,773 @@
+  // lo. Similarly, one_violation = 0001, i.e the LSB should be one, but it is
+  // 0 in lo. These make lo not satisfy the bit constraints, which results in
+  // us having to find the smallest value that satisfies bits
+  U zero_violation = lo & bits._zeros;
+  U one_violation = ~lo & bits._ones;
+  if (zero_violation == one_violation) {
+    // This means lo does not violate bits, it is the result
+    assert(zero_violation == 0, "");
+    return lo;
+  }
+
+  // The principle here is that, consider the first bit in result that is
+  // different from the corresponding bit in lo, since result is larger than lo
+  // the bit must be 0 in lo and 1 in result. As result should be the smallest
+  // value, this bit should be the rightmost one possible.
+  // E.g:      1 2 3 4 5 6
+  //      lo = 1 0 0 1 1 0
+  //       x = 1 0 1 0 1 0
+  //       y = 0 1 1 1 1 1
+  // x would be larger than lo since the first different bit is the 3rd one,
+  // while y is smaller than lo because the first different bit is the 1st bit.
+  // Next consider:
+  //      x1 = 1 0 1 0 1 0
+  //      x2 = 1 0 0 1 1 1
+  // Both x1 and x2 are larger than lo, but x1 > x2 since its first different
+  // bit from lo is the 3rd one, while with x2 it is the 7th one. As a result,
+  // if both x1 and x2 satisfy bits, x2 would be closer to our true result.
+  if (zero_violation < one_violation) {
+    // This means that the first bit that does not satisfy the bit requirement
+    // is a 0 that should be a 1, this may be the first different bit we want
+    // to find.
+    // E.g:      1 2 3 4 5 6 7 8
+    //      lo = 1 0 0 1 0 0 1 0
+    //   zeros = 0 0 1 0 0 1 0 0
+    //    ones = 0 1 0 0 1 0 1 0
+    //   1-vio = 0 1 0 0 1 0 0 0
+    //   0-vio = 0 0 0 0 0 0 0 0
+    // Since the result must have the 2nd bit set, it must be at least:
+    //           1 1 0 0 0 0 0 0
+    // This value must satisfy zeros, because all bits before the 2nd bit have
+    // already satisfied zeros, and all bits after the 2nd bit are all 0 now.
+    // Continue the logic with each 1 bit in ones, we will set each bit in our
+    // new lo (11000000 -> 11001000 -> 11001010). The final value is our
+    // result.
+    // Implementationwise, from 11000000 we can just | with ones to obtain the
+    // final result.
+
+    // first_violation is the position of the violation counting from the
+    // lowest bit up (0-based)
+    juint first_violation = W - 1 - count_leading_zeros(one_violation); // 6
+    //           0 1 0 0 0 0 0 0
+    U alignment = U(1) << first_violation;
+    // This is the first value which have the violated bit set, which means
+    // that the result should not be smaller than this
+    //           1 1 0 0 0 0 0 0
+    lo = (lo & -alignment) + alignment;
+    //           1 1 0 0 1 0 1 0
+    return lo | bits._ones;
+  }
+
+  // This is more difficult because trying to unset a bit requires us to flip
```

Comment:
Put this in an else-block, and say that
`// This means that the first bit that does not satisfy the bit requirement is a 1 that should be a 0`
But use msb rather than first.

first is always a little tricky, because it depends on big/little endian and from where you look